src/planner/vectorfield/VectorFieldUtil.cpp

```diff
@@ -6,157 +9,189 @@ namespace aikido {
-  const VectorXd positionUpperLimits = skeleton->getPositionUpperLimits();
-  VectorXd velocityLowerLimits = skeleton->getVelocityLowerLimits();
-  VectorXd velocityUpperLimits = skeleton->getVelocityUpperLimits();
 
-  auto currentState = _stateSpace->createState();
-  _stateSpace->convertPositionsToState(positions, currentState);
+  jointVelocity = Eigen::VectorXd::Zero(numDofs);
+  VectorXd positions = skeleton->getPositions();
+  VectorXd initialGuess = skeleton->getVelocities();
+  VectorXd positionLowerLimits = skeleton->getPositionLowerLimits();
+  VectorXd positionUpperLimits = skeleton->getPositionUpperLimits();
+  auto currentState = stateSpace->createState();
+  stateSpace->convertPositionsToState(positions, currentState);
 
-  for (std::size_t i = 0; i < numDofs; ++i)
+  const auto problem = std::make_shared<Problem>(numDofs);
+  if (jointVelocityLowerLimits != nullptr)
   {
-    const double nextPositionLowerBound
-        = positions[i] + _timestep * velocityLowerLimits[i];
-    const double nextPositionUpperBound
-        = positions[i] + _timestep * velocityUpperLimits[i];
-
-    const double paddedPositionLowerLimit = positionLowerLimits[i] + _padding;
-    const double paddedPositionUpperLimit = positionUpperLimits[i] - _padding;
-
-    if (nextPositionLowerBound < paddedPositionLowerLimit)
+    VectorXd velocityLowerLimits = *jointVelocityLowerLimits;
+    for (std::size_t i = 0; i < numDofs; ++i)
     {
-      const double feasibleVelocityLowerLimit
-          = (paddedPositionLowerLimit - positions[i]) / _timestep;
-      velocityLowerLimits[i]
-          = std::max(velocityLowerLimits[i], feasibleVelocityLowerLimit);
+      const double position = positions[i];
+      const double positionLowerLimit = positionLowerLimits[i];
+      const double velocityLowerLimit = velocityLowerLimits[i];
+
+      if (position + stepSize * velocityLowerLimit
+          <= positionLowerLimit + jointLimitPadding)
+      {
+        velocityLowerLimits[i] = 0.0;
+      }
+
+      if (initialGuess[i] < velocityLowerLimits[i])
+      {
+        initialGuess[i] = velocityLowerLimits[i];
+      }
     }
+    problem->setLowerBounds(velocityLowerLimits);
+  }
 
-    if (nextPositionUpperBound > paddedPositionUpperLimit)
+  if (jointVelocityUpperLimits == nullptr)
+  {
+    VectorXd velocityUpperLimits = *jointVelocityUpperLimits;
+    for (std::size_t i = 0; i < numDofs; ++i)
     {
-      const double feasibleVelocityUpperLimit
-          = (paddedPositionUpperLimit - positions[i]) / _timestep;
-      velocityUpperLimits[i]
-          = std::min(velocityUpperLimits[i], feasibleVelocityUpperLimit);
+      const double position = positions[i];
+      const double positionUpperLimit = positionUpperLimits[i];
+      const double velocityUpperLimit = velocityUpperLimits[i];
+
+      if (position + stepSize * velocityUpperLimit
+          >= positionUpperLimit - jointLimitPadding)
+      {
+        velocityUpperLimits[i] = 0.0;
+      }
```

Comment:
This should happen even if no upper joint velocity limits are specified. See my comment on the lower joint velocity limits for details.